google/resource_dataflow_job.go

```diff
@@ -0,0 +1,202 @@
+	"JOB_STATE_CANCELLED":  {},
+	"JOB_STATE_UPDATED":    {},
+	"JOB_STATE_DRAINING":   {},
+	"JOB_STATE_DRAINED":    {},
+	"JOB_STATE_CANCELLING": {},
+}
+
+func resourceDataflowJob() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDataflowJobCreate,
+		Read:   resourceDataflowJobRead,
+		Delete: resourceDataflowJobDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"template_gcs_path": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"temp_gcs_location": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"max_workers": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"parameters": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"on_delete": &schema.Schema{
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"cancel", "drain"}, false),
+				Optional:     true,
+				Default:      "drain",
+				ForceNew:     true,
+			},
+
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"state": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceDataflowJobCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	zone, err := getZone(d, config)
+	if err != nil {
+		return err
+	}
+	params := expandStringMap(d, "parameters")
+
+	env := dataflow.RuntimeEnvironment{
+		TempLocation: d.Get("temp_gcs_location").(string),
+		Zone:         zone,
+		MaxWorkers:   int64(d.Get("max_workers").(int)),
+	}
+
+	request := dataflow.CreateJobFromTemplateRequest{
+		JobName:     d.Get("name").(string),
+		GcsPath:     d.Get("template_gcs_path").(string),
+		Parameters:  params,
+		Environment: &env,
+	}
+
+	job, err := config.clientDataflow.Projects.Templates.Create(project, &request).Do()
+	if err != nil {
+		return err
+	}
+	d.SetId(job.Id)
+
+	return resourceDataflowJobRead(d, meta)
+}
+
+func resourceDataflowJobRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+
+	job, err := config.clientDataflow.Projects.Jobs.Get(project, id).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataflow job %s", id))
+	}
+
+	if _, ok := dataflowTerminalStatesMap[job.CurrentState]; ok {
+		log.Printf("[DEBUG] Removing resource '%s' because it is in state %s.\n", job.Name, job.CurrentState)
+		d.SetId("")
+		return nil
+	} else {
```

Comment:
nit: no need for this "else" (https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow)

Reply:
Ah, good point, thanks.